### PR TITLE
Update WiFi connection logic and improve Improv logging

### LINKED ISSUE #425: Can't set Wifi SSID/password for Panlee project
Followed all the instructions in flashing the ESP32-S3 dev kit (Panlee is the only project available for it it seems) and I'm not able to set my wifi SSID/password, no dialog comes up, I've reset the device, reloaded the browser, etc  (Chrome) and no dice

Here's the console output 
```
ESP-ROM:esp32s3-20210327
Build:Mar 27 2021
rst:0x1 (POWERON),boot:0x8 (SPI_FAST_FLASH_BOOT)
SPIWP:0xee
mode:DIO, clock div:1
load:0x3fce3808,len:0x44c
load:0x403c9700,len:0xbe4
load:0x403cc700,len:0x2a68
entry 0x403c98d4
E (268) esp_core_dump_flash: No core dump partition found!
E (268) esp_core_dump_flash: No core dump partition found!
Replacing Idle Tasks with TaskManager...
(I) (PrintOutputHeader)(C1) NightDriverStrip
(I) 
(I) (PrintOutputHeader)(C1) ------------------------------------------------------------------------------------------------------------
(I) (PrintOutputHeader)(C1) M5STICKC: 0, USE_M5DISPLAY: 0, USE_OLED: 0, USE_TFTSPI: 1, USE_LCD: 0, USE_AUDIO: 0, ENABLE_REMOTE: 0
(I) (PrintOutputHeader)(C1) ESP32 PSRAM Init: OK
(I) (PrintOutputHeader)(C1) Version 40: Wifi SSID: "" - ESP32 Free Memory: 287632, PSRAM:2095055, PSRAM Free: 2094659
(I) (PrintOutputHeader)(C1) ESP32 Clock Freq : 240 MHz
(I) (setup)(C1) Startup!
(I) (setup)(C1) Starting DebugLoopTaskEntry
>> Launching Debug Thread.  Mem: 287632, LargestBlk: 262132, PSRAM Free: 2094659/2095055, >> Launching JSON Writer Thread.  Mem: 279064, LargestBlk: 262132, PSRAM Free: 2094623/2095039, (W) (DeviceConfig)(C1) DeviceConfig could not be loaded from JSON, using defaults
(W) (NotifyJSONWriterThread)(C1) >> Notifying JSON Writer Thread
(W) (setup)(C1) Starting ImprovSerial
(W) (ReadWiFiConfig)(C1) Retrieved SSID and Password from NVS: , ********
(W) (setup)(C1) Creating TFT Screen
[  1057][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
E (793) gpio: gpio_set_level(226): GPIO output gpio_num error
[  1068][E][esp32-hal-gpio.c:102] __pinMode(): Invalid pin selected
E (805) gpio: gpio_set_level(226): GPIO output gpio_num error
E (816) gpio: gpio_set_level(226): GPIO output gpio_num error
E (837) gpio: gpio_set_level(226): GPIO output gpio_num error
(W) (InitializeHardware)(C1) Allocating LEDStripGFX for channel 0
(I) (AddLEDsToFastLED)(C1) Adding LEDs to FastLED...
(I) (AddLEDsToFastLED)(C1) Adding 1152 LEDs to pin 5 from channel 0 on FastLED.
(W) (SetupBufferManagers)(C1) Reserving 312 LED buffers for a total of 1088256 bytes...
(W) (InitEffectsManager)(C1) InitEffectsManager...
(I) (InitEffectsManager)(C1) Creating EffectManager using default effects
>> Launching Drawing Thread.  Mem: 265900, LargestBlk: 249844, PSRAM Free: 986607/2084863, (W) (DrawLoopTaskEntry)(C1) >> DrawLoopTaskEntry
(W) 
(W) (DrawLoopTaskEntry)(C1) Entering main draw loop!
>> Launching Screen Thread.  Mem: 261428, LargestBlk: 245748, PSRAM Free: 986607/2084863, (I) (setup)(C1) Calling ConnectToWifi()
(I) 
(I) (ConnectToWiFi)(C1) Setting host name to NightDriverStrip...WL_NO_SHIELD
(W) (ConnectToWiFi)(C1) WiFi credentials not set, cannot connect.
(W) (ConnectToWiFi)(C1) Waiting for WiFi credentials to be set...
(W) (ConnectToWiFi)(C1) WiFi credentials not set, cannot connect.
(W) (ConnectToWiFi)(C1) Waiting for WiFi credentials to be set...
(W) (ConnectToWiFi)(C1) WiFi credentials not set, cannot connect.
(W) (ConnectToWiFi)(C1) Waiting for WiFi credentials to be set...
(W) (ConnectToWiFi)(C1) WiFi credentials not set, cannot connect.
(W) (ConnectToWiFi)(C1) Waiting for WiFi credentials to be set...
```

On a side note, if you have the text size increased in your browser you can't see the whole page and there are no scroll bars.. I got that figured out pretty quick though 
image

### PR DESCRIPTION
**Pull request by rbergen**
## Description

This:

- Changes the WiFi connection logic to not block the main `setup()` function.
- Makes the stored WiFI credentials (which we consider sensitive) local to the only function that really uses them, instead of global variables that are available "everywhere".
- Maintains the possibility to demand functioning WiFi connectivity at the penalty of a reboot if it's down for too long.
- Makes the hostname a device config setting, and makes the hostname consistent across use cases (WiFi, OTA, RemoteDebug).
- Significantly improves the logging of Improv traffic to SPIFFS when enabled, to the point it is actually usable to analyze issues with Improv processing.
- Fixes #425.
- Makes index.html link to index.js instead of index.js.gz, to prevent fetch errors for the latter.
- Includes a small refactor of CWebServer to improve readability and consistency.

@KeiranHines While testing the hostname setting, I found that the web UI currently does not save (i.e. send to the /settings endpoint) empty strings. Which makes sense in all current cases, except for the hostname - one may explicitly want that to be the Espressif platform's default. I've added an optional "emptyAllowed" property to the `SettingSpec` class's JSON to indicate when a string value may be empty.  

## Contributing requirements

* [x] I read the contribution guidelines in [CONTRIBUTING.md](http://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/CONTRIBUTING.md).
* [x] I understand the BlinkenPerBit metric, and maximized it in this PR.
* [x] I selected `main` as the target branch.
* [x] All code herein is subjected to the license terms in [COPYING.txt](http://github.com/PlummersSoftwareLLC/NightDriverStrip/blob/main/COPYING.txt).

### CODE REVIEW

File: include/globals.h [hunk 1]
```diff
@@ -710,9 +710,13 @@ extern RemoteDebug Debug;           // Let everyone in the project know about it
     #define PROJECT_NAME            "Ledstrip"
     #endif
 
+    #ifndef ENABLE_WEBSERVER
     #define ENABLE_WEBSERVER            0   // Turn on the internal webserver
+    #endif
```
Comment (rbergen) [inline include/globals.h:715]: This allows ENABLE_WEBSERVER to be set from the command-line for this project, which is useful for certain debugging scenarios.

File: include/improvserial.h [hunk 2]
```diff
@@ -154,19 +147,23 @@ class ImprovSerial
...
+
             auto file = SPIFFS.open(IMPROV_LOG_FILE, FILE_APPEND);
             va_list args;
 
             va_start(args, format);
 
             // We shifted the printf parameter check to ourselves, so we can suppress the warning here
             #pragma GCC diagnostic ignored "-Wformat-nonliteral"
-            file.printf(format, args);
+            vsnprintf(lineBuffer, bufferSize, format, args);
```
Comment (rbergen) [inline include/improvserial.h:160]: This change (and those leading and trailing it) actually make the arguments after the format string get through to the log lines that are written.

File: src/network.cpp [hunk 3]
```diff
@@ -212,80 +212,93 @@ void IRAM_ATTR RemoteLoopEntry(void *)
...
     #define WIFI_WAIT_INCREASE  1000    // Increase of WiFi waiting time per cycle, in ms
+    #define WIFI_WAIT_MAX       60000   // Maximum gap between retries, in ms
 
+    #define WIFI_WAIT_INIT      (WIFI_WAIT_BASE - WIFI_WAIT_INCREASE)
 
-    bool ConnectToWiFi(uint cRetries, bool waitForCredentials = false)
+    // ConnectToWiFi
+    //
+    // Try to connect to WiFi using the SSID and password passed as arguments
+    WiFiConnectResult ConnectToWiFi(const String& ssid, const String& password)
```
Comment (rbergen) [inline src/network.cpp:226]: This overload prevents the function's user from having to deal with the oddness of having to pass pointers to the overload that does the actual work.

File: src/webserver.cpp [hunk 4]
```diff
@@ -31,6 +31,7 @@
 #include "globals.h"
 #include "webserver.h"
 #include "systemcontainer.h"
+#include "improvserial.h"
```
Comment (rbergen) [inline src/webserver.cpp:34]: We have to include this to make sure `ENABLE_IMPROV_LOGGING` is available to us.

File: config/web_projects.json [hunk 5]
```diff
@@ -35,9 +35,9 @@
             ]
         },
         {
-            "tag": "esp32esp32-s3",
+            "tag": "esp32-s3",
```
[the remark this answers is not in the document]
Comment (rbergen) [reply config/web_projects.json:38]: No, this is just an internal identifier, that being internal to our own logic to build and run the web installer. ESP Web Tools or Improv never get to see this.

File: include/improvserial.h (hunk 2 above)
Comment (rbergen) [reply include/improvserial.h:160]: No, they are not. Good catch, thanks! I will include their removal with a change I'll add to make the hostname a device setting. After testing with 3 different devices on different APs I now have 4 IP addresses assigned to different "Something"s, and it just doesn't make sense for a code-level change to be the only way to configure it.

File: src/network.cpp [hunk 6]
```diff
@@ -212,80 +212,93 @@ void IRAM_ATTR RemoteLoopEntry(void *)
...
+        else if (bPreviousConnection && WiFi.isConnected())
+        {
+            return WiFiConnectResult::Connected;
+        }
 
-        for (uint iPass = 0; iPass < cRetries; iPass++)
+        // (Re)connect if credentials have changed, or our last attempt was long enough ago
+        if (haveNewCredentials || millisAtLastAttempt == 0 || millis() - millisAtLastAttempt >= retryDelay)
         {
+            millisAtLastAttempt = millis();
```
[the remark this answers is not in the document]
Comment (rbergen) [reply src/network.cpp:262]: Not really, but because both numbers in the subtraction are unsigned long, I think it should still work.

**Comment by KeiranHines** [comment]
> Is this something you can pick up in a future update to the web UI? If so, would you like me to open an Issue for it?

Please open an issue that way it's easier to track and prioritise.

**Comment by rbergen** [comment]
@KeiranHines 

> Please open an issue that way it's easier to track and prioritise.

Makes sense. Issue number #472 it is.

**Comment by rbergen** [comment]
Merging after testing on Mesmerizer, M5StickC Plus/Spectrum and S3/Panlee.

**Pull request merged**